### PR TITLE
fix: redirect http to https but port not change

### LINKED ISSUE #7011: bug: redirect http to https but port not change
### Current Behavior

Recently, I upgraded apisix with the latest code, and found that there was an error in converting http to https. When redirecting, I always bring a port 80. It should be caused by this change：https://github.com/apache/apisix/pull/6686/files#r867816800

Maybe `set $var_x_forwarded_port $server_port;` in ngx_tpl.lua conflicts with `local ret_port = tonumber(ctx.var["var_x_forwarded_port"])` in redirect.lua

### Expected Behavior

port changed from 80 to 443

### Error Logs

_No response_

### Steps to Reproduce

add redirect and set http_to_https to true

### Environment

- APISIX version (run `apisix version`): master
- Operating system (run `uname -a`): 
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`):
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`):
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`):

### PR DESCRIPTION
**Pull request by jwrookie**
### Description

When http redirects to htpps, we use a correct port

Fixes #7011 

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: conf/config-default.yaml [hunk 1]
```diff
@@ -470,3 +470,4 @@ plugin_attr:
       connect: 60s
       read: 60s
       send: 60s
+  redirect_https_port: 8443   # the default port for use by HTTP redirects to HTTPS
```
Comment (spacewander) [inline conf/config-default.yaml:473]: We should use a structure like this:
```
redirect:
    https_port: ...
```
and comment it out as nobody wants to be redirected to 8443 by default.

File: apisix/plugins/redirect.lua [hunk 2]
```diff
@@ -148,7 +149,30 @@ function _M.rewrite(conf, ctx)
     core.log.info("plugin rewrite phase, conf: ", core.json.delay_encode(conf))
 
     local ret_code = conf.ret_code
-    local ret_port = tonumber(ctx.var["var_x_forwarded_port"])
+    local local_conf = core.config.local_conf()
+    local ret_port = core.table.try_read_attr(local_conf,
```
Comment (spacewander) [inline apisix/plugins/redirect.lua:153]: There is a wrapper to get the attribute of a plugin:
https://github.com/apache/apisix/blob/6ebe02797fb564c84b30df1865adef7f473880bf/apisix/plugins/prometheus.lua#L58

We need to use it instead.

File: t/plugin/redirect.t [hunk 3]
```diff
@@ -428,66 +428,69 @@ passed
...
+Location: https://foo.com:9443/hello
 
 
 
-=== TEST 22: redirect(pass invalid non-number to x-forwarded-port)
+=== TEST 21: redirect(port using `https default port`)
+--- yaml_config
+apisix:
+    ssl:
+        enable: null
```
Comment (spacewander) [inline t/plugin/redirect.t:479]: Let's add two more test:
1. use listen_port
2. use listen and get port in the array

File: docs/en/latest/plugins/redirect.md [hunk 4]
```diff
@@ -45,7 +45,10 @@ The `redirect` Plugin can be used to configure redirects.
 
 Only one of `http_to_https`, `uri` and `regex_uri` can be configured.
 
-* When enabling `http_to_https`, the port in the redirect URL will be the value of header `X-Forwarded-Port` or the port of the server.
+* When enabling `http_to_https`, the ports in the redirect URL will pick a value in the following order (in descending order of priority)
+  * Read `plugin_attr.redirect_https_port` from the configuration file.
```
Comment (spacewander) [inline docs/en/latest/plugins/redirect.md:49]:
```suggestion
  * Read `plugin_attr.redirect_https_port` from the configuration file (`conf/config.yaml`).
```

File: docs/zh/latest/plugins/redirect.md [hunk 5]
```diff
@@ -45,7 +45,10 @@ description: 本文介绍了关于 Apache APISIX `redirect` 插件的基本信
 
 `http_to_https`、`uri` 和 `regex_uri` 只能配置其中一个属性。
 
-* 当开启 `http_to_https` 时，重定向 URL 中的端口将是 `X-Forwarded-Port` 请求头的值或服务器的端口。
+* 当开启 `http_to_https` 时，重定向 URL 中的端口将按如下顺序选取一个值（按优先级从高到低排列）
+  * 从配置文件中读取 `plugin_attr.redirect_https_port`。
```
Comment (spacewander) [inline docs/zh/latest/plugins/redirect.md:49]:
```suggestion
  * 从配置文件（`conf/config.yaml`）中读取 `plugin_attr.redirect_https_port`。
```

File: conf/config-default.yaml (hunk 1 above)
Comment (jwrookie) [reply conf/config-default.yaml]: done

File: apisix/plugins/redirect.lua (hunk 2 above)
Comment (jwrookie) [reply apisix/plugins/redirect.lua]: done

File: t/plugin/redirect.t (hunk 3 above)
Comment (jwrookie) [reply t/plugin/redirect.t:500]: Done

File: t/plugin/redirect.t [hunk 6]
```diff
@@ -428,59 +428,83 @@ passed
 
 
 
-=== TEST 18: redirect
+=== TEST 18: redirect(port using `plugin_attr.redirect_https_port`)
```
Comment (spacewander) [inline t/plugin/redirect.t:431]: ditto

**Review by spacewander — DISMISSED** [review]

File: docs/en/latest/plugins/redirect.md [hunk 7]
```diff
@@ -45,7 +45,10 @@ The `redirect` Plugin can be used to configure redirects.
 
 Only one of `http_to_https`, `uri` and `regex_uri` can be configured.
 
-* When enabling `http_to_https`, the port in the redirect URL will be the value of header `X-Forwarded-Port` or the port of the server.
+* When enabling `http_to_https`, the ports in the redirect URL will pick a value in the following order (in descending order of priority)
+  * Read `plugin_attr.redirect_https_port` from the configuration file (`conf/config.yaml`).
+  * If `apisix.ssl` is enabling, read `apisix.ssl.listen_port` first, and if it does not exist, read `apisix.ssl.listen` and select a random port from it.
```
Comment (tokers) [inline docs/en/latest/plugins/redirect.md:50]:
```suggestion
  * If `apisix.ssl` is enabled, read `apisix.ssl.listen_port` first, and if it does not exist, read `apisix.ssl.listen` and select a port randomly from it.
```

File: docs/en/latest/plugins/redirect.md [hunk 8]
```diff
@@ -45,7 +45,10 @@ The `redirect` Plugin can be used to configure redirects.
 
 Only one of `http_to_https`, `uri` and `regex_uri` can be configured.
 
-* When enabling `http_to_https`, the port in the redirect URL will be the value of header `X-Forwarded-Port` or the port of the server.
+* When enabling `http_to_https`, the ports in the redirect URL will pick a value in the following order (in descending order of priority)
+  * Read `plugin_attr.redirect_https_port` from the configuration file (`conf/config.yaml`).
```
[the remark this answers is not in the document]
Comment (jwrookie) [reply docs/en/latest/plugins/redirect.md]: Done

File: docs/zh/latest/plugins/redirect.md [hunk 9]
```diff
@@ -45,7 +45,10 @@ description: 本文介绍了关于 Apache APISIX `redirect` 插件的基本信
 
 `http_to_https`、`uri` 和 `regex_uri` 只能配置其中一个属性。
 
-* 当开启 `http_to_https` 时，重定向 URL 中的端口将是 `X-Forwarded-Port` 请求头的值或服务器的端口。
+* 当开启 `http_to_https` 时，重定向 URL 中的端口将按如下顺序选取一个值（按优先级从高到低排列）
+  * 从配置文件（`conf/config.yaml`）中读取 `plugin_attr.redirect_https_port`。
```
[the remark this answers is not in the document]
Comment (jwrookie) [reply docs/zh/latest/plugins/redirect.md]: done

File: t/plugin/redirect.t (hunk 6 above)
Comment (jwrookie) [reply t/plugin/redirect.t]: done

**Review by spacewander — APPROVED** [review]

File: apisix/plugins/redirect.lua [hunk 10]
```diff
@@ -148,7 +150,30 @@ function _M.rewrite(conf, ctx)
...
+                    local idx = math_random(1, #ret_ports)
+                    ret_port = ret_ports[idx]
+                    if type(ret_port) == "table" then
+                        ret_port = ret_port.port
+                    end
+                end
+            end
+        end
+    end
+
```
Comment (soulbird) [inline apisix/plugins/redirect.lua:176]: Personally, I feel that this code is too inelegant. Suggest:

```lua
local function get_port(attr)
    local port
    if attr then
        port = attr.https.port
    end

    if port then
        return port
    end

    local local_conf = core.config.local_conf()
    local ssl = core.table.try_read_attr(local_conf, "apisix", "ssl")
    if not ssl or ssl["enable"] then
        return port
    end

    port = ssl["listen_port"]
    if port then
        return port
    end

    local ports = ssl["listen"]
    if ports and #ports > 0 then
        local idx = math_random(1, #ports)
        port = ret_ports[idx]
        if type(port) == "table" then
            port = port.port
        end
    end

    return port
end

local ret_port = get_port(attr)
```

File: apisix/plugins/redirect.lua [hunk 11]
```diff
@@ -148,7 +150,30 @@ function _M.rewrite(conf, ctx)
     core.log.info("plugin rewrite phase, conf: ", core.json.delay_encode(conf))
 
     local ret_code = conf.ret_code
-    local ret_port = tonumber(ctx.var["var_x_forwarded_port"])
+    local ret_port = nil
+    local attr = plugin.plugin_attr(plugin_name)
```
Comment (tokers) [inline apisix/plugins/redirect.lua:154]: Where did you define the variable `plugin_name`?

Comment (jwrookie) [reply apisix/plugins/redirect.lua:154]: here:
https://github.com/apache/apisix/blob/6ebe02797fb564c84b30df1865adef7f473880bf/apisix/plugins/redirect.lua#L65-L67

Comment (jwrookie) [reply apisix/plugins/redirect.lua:154]: Was defined before :)

File: apisix/plugins/redirect.lua (hunk 10 above)
Comment (jwrookie) [reply apisix/plugins/redirect.lua:177]: done

**Comment by spacewander** [comment]
Let's merge master to make CI pass.

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**